File: backend/app/utils/ssn.py
import re

_SSN_PATTERN = re.compile(r"^\d{3}-\d{2}-\d{4}$")


def normalize_ssn(value: str | None) -> str | None:
    if value is None:
        return None

    trimmed = value.strip()
    if not trimmed:
        return None

    digits = re.sub(r"\D", "", trimmed)
    if len(digits) != 9:
        raise ValueError("SSN must be 9 digits")

    return f"{digits[:3]}-{digits[3:5]}-{digits[5:]}"


def mask_ssn(value: str | None) -> str | None:
    if not value:
        return None

    digits = re.sub(r"\D", "", value)
    if len(digits) >= 4:
        return f"***-**-{digits[-4:]}"
    return None


def is_valid_ssn(value: str | None) -> bool:
    if value is None:
        return True
    try:
        normalized = normalize_ssn(value)
    except ValueError:
        return False
    return normalized is None or bool(_SSN_PATTERN.match(normalized))

File: frontend/src/utils/ssn.js
export function formatSsnInput(value) {
  const digits = String(value || '').replace(/\D/g, '').slice(0, 9)

  if (digits.length <= 3) return digits
  if (digits.length <= 5) return `${digits.slice(0, 3)}-${digits.slice(3)}`
  return `${digits.slice(0, 3)}-${digits.slice(3, 5)}-${digits.slice(5)}`
}

export function maskSsn(value) {
  const digits = String(value || '').replace(/\D/g, '')
  if (digits.length < 4) return '-'
  return `***-**-${digits.slice(-4)}`
}
